chain/network/src/raw.rs

```diff
@@ -216,115 +213,42 @@ impl Peer {
         Ok((len as usize, first_byte_time.unwrap_or(Instant::now())))
     }
 
-    // drains anything still available to be read without blocking and
-    // returns a vec of message lengths currently ready
-    fn read_remaining_messages(
-        &mut self,
-        first_msg_length: usize,
-    ) -> io::Result<Vec<(usize, Instant)>> {
-        let mut pos = first_msg_length + 4;
-        let mut lengths = Vec::new();
-        let mut pending_msg_length = None;
-
-        loop {
-            match self.stream.try_read_buf(&mut self.buf) {
-                Ok(n) => {
-                    if n == 0 {
-                        return Err(io::Error::new(
-                            io::ErrorKind::UnexpectedEof,
-                            "no more bytes available, but expected to receive a message",
-                        ));
-                    }
-                    let timestamp = Instant::now();
-                    tracing::trace!(target: "network", "Read {} bytes from {:?} non-blocking", n, self.stream.peer_addr());
-                    loop {
-                        if pending_msg_length.is_none() && self.buf.remaining() - pos >= 4 {
-                            let len =
-                                u32::from_le_bytes(self.buf[pos..pos + 4].try_into().unwrap());
-                            pending_msg_length = Some(len as usize);
-                        }
-                        if let Some(l) = pending_msg_length {
-                            if self.buf.remaining() - pos >= l + 4 {
-                                lengths.push((l, timestamp));
-                                pos += l + 4;
-                                pending_msg_length = None;
-                            } else {
-                                break;
-                            }
-                        } else {
-                            break;
-                        }
-                    }
-                }
-                Err(e) => match e.kind() {
-                    io::ErrorKind::WouldBlock => {
-                        break;
-                    }
-                    _ => return Err(e),
-                },
-            };
-        }
-        Ok(lengths)
-    }
-
     // Reads from the socket until there is at least one full PeerMessage available.
-    // After that point, continues to read any bytes available to be read without blocking
-    // and appends any extra messages to the returned Vec. Usually there is only one in there
-    async fn recv_messages(&mut self) -> io::Result<Vec<(PeerMessage, Instant)>> {
-        let mut messages = Vec::new();
+    async fn recv_message(&mut self) -> io::Result<(PeerMessage, Instant)> {
         let (msg_length, first_byte_time) = self.read_msg_length().await?;
 
         while self.buf.remaining() < msg_length + 4 {
-            // TODO: measure time to last byte here
             self.do_read().await?;
```

Comment:
why a loop, when you can just do the same as what you have written in the test below?:
```
 let len = read.read_u32_le().await.unwrap();
            let mut buf = vec![0; len as usize];
            read.read_exact(&mut buf).await.unwrap();
```
?

Reply:
as I mentioned above, the pattern you suggest here would make `recv()` incorrect to use in a loop with `tokio::select` as is done in this PR, since the function can be stopped and restarted in between those two awaits. So then we could still try

```rust
        self.stream.read_exact(&mut self.buf[..4]).await?;
        let len = u32::from_le_bytes(self.buf[..4].try_into().unwrap());
        self.buf.reserve(len as usize);
        self.stream.read_exact(&mut self.buf[..len]).await?;
```

but we are not using a `tokio::io::BufReader`, so the initial read of 4 bytes is not great, and the current loop has us read a larger amount into the buffer at a time. When you unravel the above to get something more efficient without a `tokio::io::BufReader`, it's not a whole lot simpler than what's here.

so then we could definitely use a `tokio::io::BufReader` and use `read_exact()`. But then we have a similar issue where we need to keep the result around between any `await`s, so feels to me just much more direct to trade a little bit of simplicity for explicit control over when we `advance()` the buffer

Reply:
Right, I missed your response there, sorry. So yes, you are right, to get cancellation safety you need a bit more that I suggested and it's fine if you want it implemented it this way.